### ISSUE TITLE
Page - Event Receiver Requirements

### ISSUE THREAD

**Issue by james-answer** [nhsconnect/Events-Management#19]
The Event Receiver Requirements page (https://developer.nhs.uk/apis/ems-beta/receiver_requirements.html) could do with some additiona information and specific points to make the requirements clearer:

1. There should be some wording to explain at this stage the NEMS does not handle errors experienced by subscribers and there is no error response mechanism back to NEMS therefor it is the responsibility of the subscriber to handling and manage errors in data or event content. Currently the spec implies that NEMS can have errors sent back to it, scentance "A key use-case for this is to allow for both positive aknowledgements and errors to be returned to the Events Management Service.".

2. We should probably also highlight the limitation of mesh and link to the specification, such as once a message is read from the mesh it can not be re-sent by NEMS so it is the responsibility of the consumer to maintain a copy or persist the message in the event of an error. Also worth pointing out that messages only stay in the MESH mailbox for 30 days so messages should be processed in a timely manner.

3. This page could do with an example of what the message recieved in the Mesh mailbox will look like.

**Comment by fletch1975** [comment]
Agree - decisions on use of ITK3 wrapper and lack of acknowledgements have only recently been made, documentation needs to be updated to reflect this.

**Comment by james-answer** [comment]
Looking more closely at the ITK3 wrapper requirements, since the EMS itself will add the ITK wrapper we should probably call out some specifics about the ITK wrapper resources the subscriber will recieve so they know what to look for.

For example, in the MessageHeader resource there is also an event type, currently the fixed valueset (https://fhir.nhs.uk/STU3/CodeSystem/ITK-MessageEvent-2) has some very specific values which are probably not appropriate for our use, are we going to have a new event type added to the value set, something like "ITK NEMS Event". Information like this would help the consumer filter out the messages that hit their mailbox.

In the "ITK-MessageHeader-2" resource there is a mandatory complex extension "ITKMessageHandling" which contains some mandatory elements, specifically the "RecipientType" extension has value of either "For Action" or "For Information", we probably should specify what it will be so that the consumer knows what they are going to recieve.

We should probably add information around the MESH control file we will be sending and how the subscribers will be able to identify the NEMS messages from all the other messages within their MESH mailbox, e.g. we need to specify the workflow ID.

**Comment by bushra** [comment]
Agreed, the requirements around the use of MessageDefinitions for the event messages will need to be confirmed, as they are a mandatory requirement in the ITK3 wrapper.

**Comment by bushra** [comment]
The Event Receiver Requirements guidance has been updated: https://developer.nhs.uk/apis/ems-beta/receiver_requirements.html 

Assigning @james-answer to oversee a review of this content with @ahatherly and @fletch1975

**Comment by james-answer** [comment]
My initial comments are as per below, but I still need to 
- validate it against the ITK3 Messaging specification
- look more closely at the example
- look at the message definition


https://developer.nhs.uk/apis/ems-beta/receiver_requirements.html

1. Needs to mention Mesh as the ITK wrapper is focused on MESH, if we used a different delivery method it might not need the ITK wrapper. This should include additional information around the workflowID that will be in the MESH control file for each message. Also limitations of MESH such as the deletion of messages after 30 days.

2. Change wording a bit at the bootom of the reciever requirements page to read better and highlight the benefits of including the wrapper. Maybe address more what it could be used for if required and bring the important warning into the main text, bullet pointing out the elements which will have fixed values at the moment. If we did expand the use in the future we would need to say what each element will be populated with and when, plus what it means so it would probably be fine to seperate all the elements with fixed values now just saying what the fixed value will be.


https://developer.nhs.uk/apis/ems-beta/itk_ems_profiles.html

3. ITK-Message-Bundle-1
- The 'entry' should probably have a cardinality of 2..* as the Message header and bundle are mandatory. Maybe we should make clear that these two resources will always be included.


4 ITK-MessageHeader-2
- The 'event' element only has an example, I assume this is going to be a fixed value for all ITK EMS event messages as the event holds the type of message so it does not need to be duplicated in this element. If this is going to be a fixed value what do we need to do to get this confirmed and put in the spec?
- The 'focus' element should probably refer to the EMS event message bundle rather than using the words "actual content" as it seems wrong. Maybe something like "The EMS-Bundle-1 origionating from the event publisher, containing the EMS event".
- The 'focus' should probably have a cardinality of 1..1 as we are expecting it to always have a event message bundle in the ITK wrapper.


5. Extension-ITK-MessageHandling-2

- Rather than using the words "represented as `false`" it might be better having more wording like "Currently EMS will only send the element with a fixed value of `false`". The current wording does not make clear that EMS will currently only send the value false. It might be better to use the word "EMS" rather than "sender" as it is the EMS which adds the ITK wrapper and using the word sender might confuse the consumer making them think the publishing system can ask for acknowledgement. The same applies for fixed values in the rest of the table.

**Comment by bushra** [comment]
> @bushra - My initial comments are as per below, but I still need to
> 
> * validate it against the ITK3 Messaging specification
> * look more closely at the example
> * look at the message definition
> 
> https://developer.nhs.uk/apis/ems-beta/receiver_requirements.html
> 
> 1. Needs to mention Mesh as the ITK wrapper is focused on MESH, if we used a different delivery method it might not need the ITK wrapper. This should include additional information around the workflowID that will be in the MESH control file for each message. Also limitations of MESH such as the deletion of messages after 30 days.
> 2. Change wording a bit at the bootom of the reciever requirements page to read better and highlight the benefits of including the wrapper. Maybe address more what it could be used for if required and bring the important warning into the main text, bullet pointing out the elements which will have fixed values at the moment. If we did expand the use in the future we would need to say what each element will be populated with and when, plus what it means so it would probably be fine to seperate all the elements with fixed values now just saying what the fixed value will be.
> 
> https://developer.nhs.uk/apis/ems-beta/itk_ems_profiles.html
> 
> 1. ITK-Message-Bundle-1
> 
> * The 'entry' should probably have a cardinality of 2..* as the Message header and bundle are mandatory. Maybe we should make clear that these two resources will always be included.
> 
> 1. ITK-MessageHeader-2
> 
> * The 'event' element only has an example, I assume this is going to be a fixed value for all ITK EMS event messages as the event holds the type of message so it does not need to be duplicated in this element. If this is going to be a fixed value what do we need to do to get this confirmed and put in the spec?
> * The 'focus' element should probably refer to the EMS event message bundle rather than using the words "actual content" as it seems wrong. Maybe something like "The EMS-Bundle-1 origionating from the event publisher, containing the EMS event".
> * The 'focus' should probably have a cardinality of 1..1 as we are expecting it to always have a event message bundle in the ITK wrapper.
> 
> 1. Extension-ITK-MessageHandling-2
> 
> * Rather than using the words "represented as `false`" it might be better having more wording like "Currently EMS will only send the element with a fixed value of `false`". The current wording does not make clear that EMS will currently only send the value false. It might be better to use the word "EMS" rather than "sender" as it is the EMS which adds the ITK wrapper and using the word sender might confuse the consumer making them think the publishing system can ask for acknowledgement. The same applies for fixed values in the rest of the table.

**Comment by james-answer** [comment]
@bushra For 1 & 2 I have updated the page on the "feature/itk_wrapper_changes" branch with changed wording to include information around MESH and more clean wording at the bottom. In the Acknowledgements and Errors section I removed the wording which states that the ITK wrapper will always specify NO acknowledgments as in the same scentance we were saying that they should implement the functionality just incase we start saying YES to acknowledgments. The new wording simplifies it to just say we want you to implement it which means when we say no they will not send responses and we can start saying yes and we should get response. I took it out of the important box as it is simply a requirement in the spec and is not more important or less important than the other requirements.
